lib/bolt/task.rb

```diff
@@ -118,5 +121,33 @@ def select_implementation(target, provided_features = [])
 
       impl
     end
+
+    def eql?(other)
+      self.class == other.class &&
+        @name == other.name &&
+        @metadata == other.metadata &&
+        @files == other.files &&
+        @remote == other.remote
+    end
+
+    alias == :eql?
+
+    def to_h
+      {
+        name: @name,
+        files: @files,
+        metadata: @metadata
+      }
+    end
+
+    def validate_metadata
```

Comment:
I think it'd be useful to have a test to verify that this warns with extra keys present - https://github.com/puppetlabs/bolt/blob/master/spec/integration/parsing_spec.rb seems like a good place for that? The `sample::params` task already has an extra top-level `anything` key, if we want to just use that.